### ISSUE TITLE
Timeout when starting kolibri-server if kolibri hasn't been running

### ISSUE THREAD

**Issue by benjaoming** [learningequality/kolibri-server#31]
Steps to reproduce:

1. Install kolibri
1. Stop kolibri with `sudo service kolibri stop` or similar
1. Remove the `~/.kolibri` folder such that a new database etc. has to be created
1. Run `sudo apt install kolibri-server`

Result

```
...
 kolibri-server.service - A high performance web server setup for Kolibri
   Loaded: loaded (/lib/systemd/system/kolibri-server.service; enabled; vendor preset: enabled)
   Active: failed (Result: timeout) since Thu 2019-08-22 18:57:40 BST; 67ms ago
  Process: 4684 ExecStart=/etc/init.d/kolibri-server start (code=killed, signal=TERM)
 Main PID: 3558 (code=killed, signal=KILL)
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/kolibri-server.service
           └─4710 systemctl start kolibri.service

Aug 22 18:56:10 raspberrypi su[4696]: Successful su for pi by root
Aug 22 18:56:10 raspberrypi su[4696]: + ??? root:pi
Aug 22 18:56:10 raspberrypi su[4696]: pam_unix(su:session): session opened for user pi by (uid=0)
Aug 22 18:56:12 raspberrypi kolibri-server[4684]: WARNING:root:No C Extensions available for this platform.
Aug 22 18:56:14 raspberrypi kolibri-server[4684]: WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is re…ke effect.
Aug 22 18:56:14 raspberrypi kolibri-server[4684]: debconf: DbDriver "passwords" warning: could not open /var/cache/debconf/passwords.dat: Pe…ion denied
Aug 22 18:57:40 raspberrypi systemd[1]: kolibri-server.service: Start operation timed out. Terminating.
Aug 22 18:57:40 raspberrypi systemd[1]: Failed to start A high performance web server setup for Kolibri.
Aug 22 18:57:40 raspberrypi systemd[1]: kolibri-server.service: Unit entered failed state.
Aug 22 18:57:40 raspberrypi systemd[1]: kolibri-server.service: Failed with result 'timeout'.
Hint: Some lines were ellipsized, use -l to show in full.
dpkg: error processing package kolibri-server (--configure):
 subprocess installed post-installation script returned error exit status 1
Processing triggers for systemd (232-25+deb9u11) ...
Errors were encountered while processing:
 kolibri-server
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by benjaoming** [comment]
Theory:

`sudo apt install kolibri-server` will install and perform the first run of kolibri separately. Thus, the timeout-resistent `kolibri.service` runs kolibri for the first time.

If we delete the `~/.kolibri` directory and run `kolibri-server.service`, then the init script will be invoked and start kolibri with a clean `KOLIBRI_HOME`, prompting a longer runtime and eventually the timeout.

This was on a Raspberry Pi, so potentially a slower device.

**Comment by intelliant01** [comment]
@jredrejo Have you tested the stop timeout with the large database? I have never seen a clean kolibri and kolibri-server shutdown with the large db. Kolibri is 'killed with force' and kolibri-server always reports a timeout as above. Even 10 seconds are not going to be enough as I had already tried.

**Comment by benjaoming** [comment]
I'm re-opening this, pending more feedback from @intelliant01 - it seems that 10 seconds may not be enough or there is something else causing unclean shutdowns.

I also want to verify my own setup that reproduced timeouts on RPi.

**Comment by benjaoming** [comment]
I still reproduce this issue on Raspberry Pi. Investigating the timeout.


```
root@raspberrypi:~# apt upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Calculating upgrade... Done
The following package was automatically installed and is no longer required:
  rpi.gpio-common
Use 'apt autoremove' to remove it.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
Setting up kolibri-server (0.3.5-0ubuntu1) ...
WARNING:root:No C Extensions available for this platform.

WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
Job for kolibri-server.service failed because a timeout was exceeded.
See "systemctl status kolibri-server.service" and "journalctl -xe" for details.
invoke-rc.d: initscript kolibri-server, action "start" failed.
● kolibri-server.service - A high performance web server setup for Kolibri
   Loaded: loaded (/lib/systemd/system/kolibri-server.service; enabled; vendor preset: enabled)
   Active: failed (Result: timeout) since Mon 2020-01-13 14:52:55 GMT; 54ms ago
  Process: 8848 ExecStart=/etc/init.d/kolibri-server start (code=exited, status=0/SUCCESS)
  Process: 8972 ExecStop=/etc/init.d/kolibri-server stop (code=killed, signal=TERM)

Jan 13 14:52:17 raspberrypi kolibri-server[8848]: WARNING:root:No C Extensions available for this platform.
Jan 13 14:52:18 raspberrypi kolibri-server[8848]: WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
Jan 13 14:52:19 raspberrypi kolibri-server[8848]: WARNING  Options file /home/pi/.kolibri/options.ini has been updated; server restart is required before change will take effect.
Jan 13 14:52:19 raspberrypi kolibri-server[8848]: debconf: DbDriver "passwords" warning: could not open /var/cache/debconf/passwords.dat: Permission denied
Jan 13 14:52:43 raspberrypi kolibri-server[8848]: start-stop-daemon: matching on world-writable pidfile /var/run/kolibri-server/uwsgi.pid is insecure
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Stopping timed out. Terminating.
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Control process exited, code=killed, status=15/TERM
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Killing process 8988 (systemctl) with signal SIGKILL.
Jan 13 14:52:55 raspberrypi systemd[1]: kolibri-server.service: Failed with result 'timeout'.
Jan 13 14:52:55 raspberrypi systemd[1]: Failed to start A high performance web server setup for Kolibri.
dpkg: error processing package kolibri-server (--configure):
 installed kolibri-server package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 kolibri-server
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by benjaoming** [comment]
It's not exactly the same issue, but pretty close. Starting doesn't time out, it's now stopping that times out :)

**Comment by jredrejo** [comment]
Ok, if I understand it correctly the problem is kolibri-server does not restarts correctly because kolibri does not finish correctly, should not the solution be applied to kolibri?

**Comment by benjaoming** [comment]
Do you know what part of the debian scripts that make things stop/start (or restart) during upgrade? I can't find it anywhere. But yes, we should definitely make things more robust somehow for the upgrades.

There is a `set -e` that could cause things to fail loud and clear, which is great.

**Comment by indirectlylit** [comment]
> It's not exactly the same issue, but pretty close

since this particular issue was addressed by #33 in v0.3.6, opened a new issue for v0.3.7 https://github.com/learningequality/kolibri-server/issues/51
